### ISSUE TITLE
CoBEVT在intersection场景下的训练过拟合？[Does CoBEVT overfit when training in intersection scenarios?]

### ISSUE THREAD

**Issue by jiang-zl** [AIR-THU/DAIR-RCooper#56]
I ran the **OPV2V format** training program in the Intersection scenario, completely following the configuration parameters of the repo (**The corresponding document also recommends that the OPV2V format be used to train CoBEVT in intersection scenarios**), but found that **I couldn't re-product the Table 4 in the paper：AP@0.3 : 53.5， AP@0.5 : 45.6， AP@0.7: 32.6** 

Even if I choose the model based on the optimal validation set loss, the accuracy can only reach: **AP@0.3: 40.3，AP@0.5: 32.7，AP@0.7: 24.0**

According to the given config. yml, the loss of the validation set during the training process is as follows:

At epoch 0, the validation loss is 3.298137
At epoch 2, the validation loss is 2.678899
At epoch 4, the validation loss is 3.764347
At epoch 6, the validation loss is 3.161125
At epoch 8, the validation loss is 2.913187
At epoch 10, the validation loss is 2.744693
At epoch 12, the validation loss is 2.853970
At epoch 14, the validation loss is 2.584596
At epoch 16, the validation loss is 2.996565
At epoch 18, the validation loss is 2.607438
At epoch 20, the validation loss is 2.640296
At epoch 22, the validation loss is 2.734751
At epoch 24, the validation loss is 2.371627
At epoch 26, the validation loss is 2.525078
At epoch 28, the validation loss is 2.521601
At epoch 30, the validation loss is 2.624230
At epoch 32, the validation loss is 2.781028
**At epoch 34, the validation loss is 2.985394
At epoch 36, the validation loss is 3.149842
At epoch 38, the validation loss is 3.184978
At epoch 40, the validation loss is 3.256760
At epoch 42, the validation loss is 3.446493
At epoch 44, the validation loss is 3.482708
At epoch 46, the validation loss is 3.538414
At epoch 48, the validation loss is 3.681197**

It seems that the model has been overfitting since the 35th round?

All of configurations when running the CoBEVT:
```
data_augment:
- ALONG_AXIS_LIST:
  - x
  NAME: random_world_flip
- NAME: random_world_rotation
  WORLD_ROT_ANGLE:
  - -0.78539816
  - 0.78539816
- NAME: random_world_scaling
  WORLD_SCALE_RANGE:
  - 0.95
  - 1.05
fusion:
  args: []
  core_method: IntermediateFusionDataset
loss:
  args:
    cls_weight: 1.0
    reg: 2.0
  core_method: point_pillar_loss
lr_scheduler:
  core_method: cosineannealwarm
  epoches: 50
  lr_min: 2.0e-05
  warmup_epoches: 10
  warmup_lr: 0.0002
model:
  args:
    anchor_number: 2
    backbone_fix: false
    base_bev_backbone:
      layer_nums:
      - 3
      - 5
      - 8
      layer_strides:
      - 2
      - 2
      - 2
      num_filters:
      - 64
      - 128
      - 256
      num_upsample_filter:
      - 128
      - 128
      - 128
      upsample_strides:
      - 1
      - 2
      - 4
    compression: 0
    fax_fusion:
      agent_size: 5
      depth: 3
      dim_head: 32
      drop_out: 0.1
      input_dim: 256
      mask: true
      mlp_dim: 256
      window_size: 4
    lidar_range: &id001
    - -200
    - -200
    - -8
    - 200
    - 200
    - 0
    max_cav: 5
    pillar_vfe:
      num_filters:
      - 64
      use_absolute_xyz: true
      use_norm: true
      with_distance: false
    point_pillar_scatter:
      grid_size: !!python/object/apply:numpy.core.multiarray._reconstruct
        args:
        - !!python/name:numpy.ndarray ''
        - !!python/tuple
          - 0
        - !!binary |
          Yg==
        state: !!python/tuple
        - 1
        - !!python/tuple
          - 3
        - !!python/object/apply:numpy.dtype
          args:
          - i8
          - false
          - true
          state: !!python/tuple
          - 3
          - <
          - null
          - null
          - null
          - -1
          - -1
          - 0
        - false
        - !!binary |
          IAMAAAAAAAAgAwAAAAAAAAEAAAAAAAAA
      num_features: 64
    shrink_header:
      dim:
      - 256
      input_dim: 384
      kernal_size:
      - 3
      padding:
      - 1
      stride:
      - 2
    voxel_size: &id002
    - 0.5
    - 0.5
    - 8
  core_method: point_pillar_cobevt
name: corpbevtlidar
optimizer:
  args:
    eps: 1.0e-10
    weight_decay: 0.0001
  core_method: Adam
  lr: 0.001
postprocess:
  anchor_args:
    D: 1
    H: 800
    W: 800
    cav_lidar_range: *id001
    feature_stride: 4
    h: 1.56
    l: 3.9
    num: 2
    r:
    - 0
    - 90
    vd: 8
    vh: 0.5
    vw: 0.5
    w: 1.6
  core_method: VoxelPostprocessor
  max_num: 100
  nms_thresh: 0.15
  order: hwl
  target_args:
    neg_threshold: 0.45
    pos_threshold: 0.6
    score_threshold: 0.2
preprocess:
  args:
    max_points_per_voxel: 32
    max_voxel_test: 70000
    max_voxel_train: 32000
    voxel_size: *id002
  cav_lidar_range: *id001
  core_method: SpVoxelPreprocessor
root_dir: /root/autodl-tmp/data/output/OPV2V/data/intersection/train
test_dir: /root/autodl-tmp/data/output/OPV2V/data/intersection/val
train_params:
  batch_size: 1
  epoches: 50
  eval_freq: 2
  max_cav: 5
  save_freq: 1
validate_dir: /root/autodl-tmp/data/output/OPV2V/data/intersection/val
wild_setting:
  async: false
  async_overhead: 100
  backbone_delay: 10
  data_size: 1.06
  loc_err: false
  ryp_std: 0.2
  seed: 20
  transmission_speed: 27
  xyz_std: 0.2
yaml_parser: load_point_pillar_params
```

I think the key config is summaried as below:
```
loss:
  args:
    cls_weight: 1.0
    reg: 2.0
  core_method: point_pillar_loss

lr_scheduler:
  core_method: cosineannealwarm
  epoches: 50
  lr_min: 2.0e-05
  warmup_epoches: 10
  warmup_lr: 0.0002

optimizer:
  args:
    eps: 1.0e-10
    weight_decay: 0.0001
  core_method: Adam
  lr: 0.001

train_params:
  batch_size: 1
  epoches: 50
  eval_freq: 2
  max_cav: 5
  save_freq: 1
```

**Notably**: I am using **two 32GB vGPU cards** and **conducting DDP distributed parallel training** using `torch.distributed` and `torch.nn.parallel.DistributedDataParallel`. 

Therefore, **the batch_size=1 in our config. yml (equivalent to centralized training with batch_size=2 for a single GPU card).** 

**Thank you for your amazing work, and I sincerely request your help in resolving this issue.**

**Comment by ryhnhao** [comment]
Hi, @jiang-zl, thank you for your interest in our work, and apologies for the delayed response.

We trained the model using 4×80GB A100 GPUs, so the effective batch size was significantly larger than in your setting. In your case, trying a smaller learning rate could be an effective adjustment.

Also, from the loss curve you provided, it appears the model is underfitting rather than overfitting — likely due to the learning rate being too high to allow stable convergence.

**Comment by jiang-zl** [comment]
Hello！batch size的参数我是用你们给出的config.yml来做的，我检查了你们给出的checkpoint里面的config，其中batch size为4。我用两张32G的显卡跑数据并行，每张显卡持有batch size为2的数据。在DDP下，需要配置batch size为1/2，能实现原batch size的训练。因此我们的参数应该是对齐的，如果你们训练CoBEVT时的配置参数和checkpoint不一样，好恳请您告知，感谢！


另外 关于过拟合和欠拟合，我还打印了训练的loss，是持续降低的，我明天会附加到这个issue下。而且，您说的调低学习率，根据batch size的降低倍数K，我将学习率从α/sqrt(K)到α/K都尝试过了，当然这是在做数据并行之前。


然而，我做了数据并行后，理应用你们原始的config.yml就能完成训练，请问是我哪里做的不对吗？恳请您回复、赐教！



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2025年7月7日(周一) 晚上10:51
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [AIR-THU/DAIR-RCooper] CoBEVT在intersection场景下的训练过拟合？[Does CoBEVT overfit when training in intersection scenarios?] (Issue #56)跑跑


ryhnhao left a comment (AIR-THU/DAIR-RCooper#56)
 
Hi, @jiang-zl, thank you for your interest in our work, and apologies for the delayed response.
 
We trained the model using 4×80GB A100 GPUs, so the effective batch size was significantly larger than in your setting. In your case, trying a smaller learning rate could be an effective adjustment.
 
Also, from the loss curve you provided, it appears the model is underfitting rather than overfitting — likely due to the learning rate being too high to allow stable convergence.

**Comment by jiang-zl** [comment]
Hello！batch size的参数我是用你们给出的config.yml来做的，我检查了你们给出的checkpoint里面的config，其中batch size为2。我用两张32G的显卡跑数据并行，每张显卡持有batch size为1的数据，这种配置是完全可以cover掉跑CoBEVT所需的资源。**在DDP下，需要配置进程接收的batch size为1，能实现有效batch size为`1*nprocs=2`的训练。**（[参考资料：每个进程将接收一个包含 32 个样本的输入批次；有效批次大小是 32 * nprocs，在使用 4 个 GPU 时为 128。](https://pytorch.ac.cn/tutorials/beginner/ddp_series_multigpu.html)） 因此我们的参数应该是对齐的，如果你们训练CoBEVT时的配置参数和checkpoint不一样，还恳请您告知，感谢！

另外 关于过拟合和欠拟合，我还打印了训练的loss，是持续降低的（但是loss只能降低到0.088，即小数点后两位），我已经附加到这个issue下了。而且，您说的调低学习率，根据batch size的降低倍数K，我将学习率从`α/sqrt(K)`到`α/K`都尝试过了[1]，当然这是在做数据并行之前。

然而，我做了数据并行后，理应用你们原始的config.yml就能完成训练，请问是我哪里做的不对吗？恳请您回复、赐教！

**在此，非常感谢您及您的团队做的宝贵工作，也非常感谢您的回复！** 如果DAIR-RCooper项目需要更新config.yml的配置参数，或者是我做错了什么地方，还恳请告知和赐教！

我的**训练集loss**：
At epoch 0, the training loss is 2.495526
At epoch 1, the training loss is 1.134703
At epoch 2, the training loss is 0.945408
At epoch 3, the training loss is 0.822400
At epoch 4, the training loss is 0.753850
At epoch 5, the training loss is 0.689085
At epoch 6, the training loss is 0.646770
At epoch 7, the training loss is 0.609821
At epoch 8, the training loss is 0.569043
At epoch 9, the training loss is 0.557053
At epoch 10, the training loss is 0.502330
At epoch 11, the training loss is 0.471472
At epoch 12, the training loss is 0.450925
At epoch 13, the training loss is 0.439150
At epoch 14, the training loss is 0.416362
At epoch 15, the training loss is 0.404205
At epoch 16, the training loss is 0.389526
At epoch 17, the training loss is 0.370164
At epoch 18, the training loss is 0.361705
At epoch 19, the training loss is 0.350897
At epoch 20, the training loss is 0.342709
At epoch 21, the training loss is 0.329188
At epoch 22, the training loss is 0.319913
At epoch 23, the training loss is 0.306231
At epoch 24, the training loss is 0.295316
At epoch 25, the training loss is 0.288462
At epoch 26, the training loss is 0.278129
At epoch 27, the training loss is 0.269843
At epoch 28, the training loss is 0.255855
At epoch 29, the training loss is 0.250066
At epoch 30, the training loss is 0.236496
At epoch 31, the training loss is 0.225975
At epoch 32, the training loss is 0.214118
At epoch 33, the training loss is 0.205400
At epoch 34, the training loss is 0.194362
At epoch 35, the training loss is 0.182224
At epoch 36, the training loss is 0.171592
At epoch 37, the training loss is 0.161084
At epoch 38, the training loss is 0.154106
At epoch 39, the training loss is 0.143110
At epoch 40, the training loss is 0.135043
At epoch 41, the training loss is 0.128391
At epoch 42, the training loss is 0.119969
At epoch 43, the training loss is 0.112644
At epoch 44, the training loss is 0.107041
At epoch 45, the training loss is 0.101663
At epoch 46, the training loss is 0.097468
At epoch 47, the training loss is 0.093625
At epoch 48, the training loss is 0.090653
At epoch 49, the training loss is 0.088433

参考文献：
[1] 苏剑林. (Nov. 14, 2024). 《当Batch Size增大时，学习率该如何随之变化？ 》[Blog post]. Retrieved from https://spaces.ac.cn/archives/10542

**Comment by ryhnhao** [comment]
@jiang-zl

抱歉我之前的回复可能有些误导，时间较久，对当时训练细节的回忆存在偏差。
实际上，我们训练 CoBEVT 时是应该使用了单张 GPU，配置中的 batch_size=2 是指单卡 batch size，因此你现在在两张卡上使用 batch_size=1（DDP，总 effective batch size = 2）应该是等价的配置，设置上应该没有问题。

此外，你提供的训练 loss 显示模型确实在稳定下降，之前我仅从 val loss 曲线推测欠拟合显然是不充分的分析，对此表示抱歉。这也在你后来提供train loss后应该得到了澄清。

关于训练轮数，我们在 RCooper benchmark 中为了公平对比不同模型，统一训练了 50 个 epoch，并在论文中汇报了最后一个 epoch 的结果。因此如果想对齐论文结果，按理说训练轮数和配置保持一致即可。

你有尝试在验证集上直接加载我们公开的 checkpoint 并跑一次评估吗？如果结果能够对齐，那就说明目前的差异应该并非来自硬件、数值精度或实现细节，而可能与训练动态有关。如果是这样的话，建议和我们保持一致使用单卡训练。

再次感谢你认真细致的复现与反馈，也感谢你的理解！

**Comment by jiang-zl** [comment]
好的，非常非常感谢您的帮助！！！
